server/admin/admin_server.go

```diff
@@ -82,24 +83,39 @@ func (s *Server) CreateTree(ctx context.Context, request *trillian.CreateTreeReq
 	if _, err := trees.Hasher(tree); err != nil {
 		return nil, status.Errorf(codes.InvalidArgument, "failed to create hasher for tree: %v", err.Error())
 	}
-	signer, err := trees.Signer(ctx, s.registry.SignerFactory, tree)
-	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "failed to create signer for tree: %v", err.Error())
-	}
 
-	// Derive the public key that corresponds to the private key for this tree.
-	// The caller may have provided the public key, but for safety we shouldn't rely on it being correct.
-	publicKeyDER, err := x509.MarshalPKIXPublicKey(signer.Public())
-	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "failed to marshal public key: %v", err.Error())
-	}
-	// If a public key was provided, check that it matches the one we derived. If it doesn't, this indicates a mistake by the caller.
-	if tree.PublicKey != nil && !bytes.Equal(tree.PublicKey.Der, publicKeyDER) {
-		return nil, status.Error(codes.InvalidArgument, "the public and private keys are not a pair")
-	}
-	// If no public key was provided, use the DER that we just marshaled.
-	if tree.PublicKey == nil {
-		tree.PublicKey = &trillian.PublicKey{Der: publicKeyDER}
+	// If a private key was provided, validate it. If a key was not provided, generate one.
+	switch {
+	case tree.PrivateKey != nil:
+		signer, err := trees.Signer(ctx, s.registry.SignerFactory, tree)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "failed to create signer for tree: %v", err.Error())
+		}
+
+		// Derive the public key that corresponds to the private key for this tree.
+		// The caller may have provided the public key, but for safety we shouldn't rely on it being correct.
+		publicKeyDER, err := x509.MarshalPKIXPublicKey(signer.Public())
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "failed to marshal public key: %v", err.Error())
+		}
+
+		// If a public key was provided, check that it matches the one we derived. If it doesn't, this indicates a mistake by the caller.
+		if tree.PublicKey != nil && !bytes.Equal(tree.PublicKey.Der, publicKeyDER) {
+			return nil, status.Error(codes.InvalidArgument, "the public and private keys are not a pair")
+		}
+
+		// If no public key was provided, use the DER that we just marshaled.
+		if tree.PublicKey == nil {
+			tree.PublicKey = &keyspb.PublicKey{Der: publicKeyDER}
+		}
+	case request.KeyGen != nil:
+		newTree, err := s.registry.SignerFactory.Generate(ctx, tree, request.KeyGen)
+		if err != nil {
+			return nil, status.Errorf(codes.FailedPrecondition, "failed to generate private key: %v", err.Error())
```

Comment:
`InvalidArgument` instead of `FailedPrecondition`? There isn't much system state that could interfere in key generation (at least not in the current impl, it seems).